Clubb coriolis

HingOng opened the pull request:
These changes implement the nontraditional and traditional Coriolis terms, each controlled by a flag. The nontraditional ones are explicitly integrated in the up2, wp2, and upwp equations, and so are the traditional ones in the upwp and vpwp equations. The Coriolis parameters are either passed through the host model interface or calculated from the latitude in the standalone namelist. The Coriolis terms are ignored by the default flag setting.

With the default setting, the only change is the setting of the traditional Coriolis parameter from the standalone namelist. Previous versions allow user to set a Coriolis parameter inconsistent with the latitude. The present version ignores the "fcor_nl" and calculates "fcor" from "lat_vals".

The implementation of the Coriolis terms has been validated with an analytical benchmark.

LarsonGroupSysAdmin committed:
Updating the clubb core and silhs version files

huebleruwm committed:
Fixing bug. qclvar copying is handled by the api, but we were deleting it with an unstructured data statement. I added this due as a fix before thinking qclvar was a local variable, which happened because I mixed up the CLUBB_CAM and CLUBBND_CAM flags. https://github.com/larson-group/cam/issues/175

huebleruwm committed:
Fixing GPU bug. There is a vertical dependency with wpxp. This is what was breaking the ECT test, see larson-group/cam#175

huebleruwm committed:
Moving pdf_params copying to copyin. This is the data structure that contains pointers, and these pointers will be different on the host and device, so it is bad in theory to copy the structure back to the CPU as it might overwrite cpu memory pointers with gpu memory pointers. In practice though I've seen no problems caused by this, I'm just making this commit preemptively, and it has already been tested with the ECT test.

huebleruwm committed:
Moving PosInf to clubb_model_settings.F90

LarsonGroupSysAdmin committed:
Updating the clubb core and silhs version files

HingOng committed:
Add flags for the Coriolis terms.

HingOng committed:
Pass fcory from api module into advance module

HingOng committed:
Implement the traditional Coriolis terms

HingOng committed:
Implement the nontraditional Coriolis terms

HingOng committed:
Merge remote-tracking branch 'origin/clubb_Coriolis_cam' into clubb_Coriolis

HingOng committed:
Add stats variables for the Coriolis terms

HingOng committed:
Compute Coriolis parameters from namelist latitude

HingOng committed:
Create the coriolis_test case

HingOng committed:
Tune the coriolis_test case. However, oscillation frequency is too slow

HingOng committed:
Correct a coefficient in Coriolis. "coriolis_test" matches analytical

vlarson commented on input/stats/host_budgets_stats.in:
```diff
@@ -39,15 +39,15 @@ vars_zm =
 'vp2_bt', 'vp2_ma', 'vp2_ta', 'vp2_tp', 'vp2_dp1', 'vp2_dp2', 
 'vp2_pr1', 'vp2_pr2', 'vp2_sdmp', 'vp2_cl', 'vp2_pd', 'vp2_sf', 'vp2_splat',
-'up2_bt', 'up2_ma', 'up2_ta', 'up2_tp', 'up2_dp1', 'up2_dp2', 
+'up2_bt', 'up2_ma', 'up2_ta', 'up2_tp', 'up2_dp1', 'up2_dp2', 'up2_nct',
```
We need to update the budget balance test:

https://github.com/larson-group/clubb_release/tree/master/postprocessing/check_budgets_balance

vlarson commented on src/CLUBB_core/version_clubb_core.txt:
```diff
@@ -0,0 +1,312 @@
+commit 51107487c8ae0f96461e5a882a203d3019aa6aa5 (HEAD -> clubb_Coriolis_cam, newremote.00/clubb_Coriolis_cam)
```
We won't need to commit this file because it's automatically generated.

vlarson commented on src/CLUBB_core/clubb_api_module.F90:
```diff
@@ -1402,6 +1405,10 @@ subroutine advance_clubb_core_api_single_col( gr, &       ! intent(in)
     !$acc end data
     !$acc end data
+      
+#ifdef CLUBB_CAM
```
Why is this needed?

vlarson replied:
Hing thinks that this was included by accident, but he hasn't tested the code on a GPU, so he's not 100% sure.

vlarson commented on input/tunable_parameters/coriolis_test_parameters.in:
```diff
@@ -0,0 +1,110 @@
+! $Id$
```
Why do we need this file?  Can we just add any new parameters to 

https://github.com/larson-group/clubb_release/blob/master/input/tunable_parameters/tunable_parameters.in ?

vlarson replied:
coriolis_test_parameters.in shuts off all the pressure terms so that Hing can compare with his analytic solution.

vlarson commented on src/SILHS/version_silhs.txt:
```diff
@@ -0,0 +1,288 @@
+commit 46cac2787c3802be0418762948ad6d1231dde182
```
Likewise, we won't need to commit this file because it's automatically generated.

vlarson commented on src/clubb_driver.F90:
```diff
@@ -988,7 +997,10 @@ subroutine run_clubb ( ngrdcol, calls_per_out, l_output_multi_col, l_output_doub
-      fcor_nl
+      fcor_nl ! obsolete
+      ! fcor_nl can be read from a namelist but is not used for any calculation.
+      ! fcor and fcory are calculated from lat_vals from a namelist.
```
Is fcor always generated from lat_vals?  I think we want to retain the capability to directly specify fcor.

vlarson replied:
Because of this change, we don't expect the results to be BFB.  We'll have to see how much the answers change.  If not much, then we could delete fcor_nl.  But if the answers do change, then maybe we do want to keep the option of specifying fcor, just so that we match our benchmark LES better.

vlarson commented on src/clubb_driver.F90:
```diff
@@ -1086,7 +1099,7 @@ subroutine run_clubb ( ngrdcol, calls_per_out, l_output_multi_col, l_output_doub
     sens_ht   = 0._core_rknd
     latent_ht = 0._core_rknd
-    fcor_nl   = 1.e-4_core_rknd
+    fcor_nl   = 1.e-4_core_rknd ! obsolete
```
Did you make this obsolete?

vlarson commented on src/clubb_driver.F90:
```diff
@@ -1259,7 +1274,8 @@ subroutine run_clubb ( ngrdcol, calls_per_out, l_output_multi_col, l_output_doub
     p_sfc = p_sfc_nl
     T_sfc = T_sfc_nl
-    fcor = fcor_nl
+    fcor  = two * omega_planet * sin ( lat_vals*radians_per_deg ) ! Retire fcor_nl
```
I think that we want this to be optional, if fcor_nl is not specified.

HingOng replied:
I have run a cgils_s6 test with this line restored as "fcor = fcor_nl" and compared the results against those from the "larson-group:master" branch. They match BFB.

A problem is that cgils_s6 test is extremely sensitive to fcor.
If prescribed from fcor_nl, fcor =   4.264000000000000E-005, the cgils_s6 test completes normally.
If calculated from lat_vals, fcor =   4.263948942216714E-005, the cgils_s6 test aborts with the following error:
iteration =      118; time =    35400.0
    Fatal error in CLUBB - MPI Process   1 / Chunk index   1
Column index not av
 ailable. Latitude/Longitude for column 1:  17.00/******
Latitude range from -90
  to 90 degrees, Longitude range from -180 to 180 degrees
 mixt_frac * precip_frac_1 + ( 1 - mixt_frac ) * precip_frac_2 /= precip_frac wi
 thin numerical roundoff
 level =           15
 mixt_frac * precip_frac_1 + ( 1 - mixt_frac ) * precip_frac_2 =
  8.254983776519802E-002
 precip_frac =   8.254983776519802E-002
 Fatal error in CLUBB in - MPI Process   1 / Chunk index   1
Grid column index i
  =  1, Latitude  17.00 / Longitude  17.00
Latitude range from -90 to 90 degrees
 , Longitude range from -180 to 180 degrees
 precip_frac_assert_check failed
    Fatal error in CLUBB - MPI Process   1 / Chunk index   1
Column index not av
 ailable. Latitude/Longitude for column 1:  17.00/******
Latitude range from -90
  to 90 degrees, Longitude range from -180 to 180 degrees
  in setup_pdf_parameters after calling precip_fraction
    Fatal error in CLUBB - MPI Process   1 / Chunk index   1
Column index not av
 ailable. Latitude/Longitude for column 1:  17.00/******
Latitude range from -90
  to 90 degrees, Longitude range from -180 to 180 degrees
 Fatal error after setup_pdf_parameters_api
    Fatal error in CLUBB - MPI Process   1 / Chunk index   1
Column index not av
 ailable. Latitude/Longitude for column 1:  17.00/******
Latitude range from -90
  to 90 degrees, Longitude range from -180 to 180 degrees
 Fatal error after pdf_hydromet_microphys_prep
CLUBB-TIMER time_loop_init =             0.0022
CLUBB-TIMER time_clubb_advance =         0.0456
CLUBB-TIMER time_clubb_pdf =             0.0283
CLUBB-TIMER time_SILHS =                 0.0001
CLUBB-TIMER time_microphys_scheme =      0.0103
CLUBB-TIMER time_microphys_advance =     0.0171
CLUBB-TIMER time_loop_end =              0.1028
CLUBB-TIMER time_output_multi_col =      0.0001
CLUBB-TIMER time_total =                 0.2073
Fatal error in clubb, check your parameter values and timestep

I had to reduce the timestep to normally complete the run with calculated fcor.

vlarson commented on src/clubb_driver.F90:
```diff
@@ -2631,7 +2651,7 @@ subroutine run_clubb ( ngrdcol, calls_per_out, l_output_multi_col, l_output_doub
     do k = 1, gr%nzm
       do i = 1, ngrdcol
-        upwp(i,k)             = zero          ! vertical u momentum flux
+        upwp(i,k)             = upwp_init(k)  ! vertical u momentum flux
```
Is upwp_init initialized for all cases?  Usually, it should be zero, no?

vlarson reviewed (changes requested):
Is this pull request a superset of https://github.com/larson-group/clubb_release/pull/20 ?

Do these code changes leave the results BFB identical when the flags are set to zero?

I can make these code changes if you don't have time or interest.

vlarson commented:
> Is this pull request a superset of #20 ?

No, #20 is a pull request for the CAM tag.  #19 is a pull request for CLUBB master.

> Do these code changes leave the results BFB identical when the flags are set to zero?

Hing hasn't done this test.

HingOng committed:
Update advance_clubb_core_module.F90

Rename fcory as fcor_y and add description

HingOng committed:
Update advance_wp2_wp3_module.F90

Rename fcory as fcor_y and add description

HingOng committed:
Update advance_xm_wpxp_module.F90

Rename fcory as fcor_y and add description

HingOng committed:
Update advance_xp2_xpyp_module.F90

Rename fcory as fcor_y and add description

HingOng committed:
Update clubb_api_module.F90

Rename fcory as fcor_y and add description

HingOng committed:
Update spurious_source_test.F90

Rename fcory as fcor_y and add description

HingOng committed:
Update clubb_driver.F90

Rename fcor as fcor_y and add description.
Also, add description for the coriolis_test

HingOng committed:
Update generalized_grid_test.F90

Rename fcory as fcor_y and add description.

HingOng committed:
Update coriolis_test_model.in

Add description for the Coriolis test case

HingOng committed:
Stablize the cgils_s6 case with smaller time step

HingOng committed:
Create the ekman case

HingOng committed:
Tune the ekman case to match LES

HingOng committed:
Compute wp2up with PDF and pass along with wp2thvp

HingOng committed:
Implement the nontraditional Coriolis term for wp3

HingOng committed:
Fine-tune parameters for the ekman case

HingOng committed:
Create the atex_long case

HingOng committed:
Retune ekman case with l_vary_convect_depth = .true.

HingOng committed:
Fine-tune the atex_long case

HingOng committed:
Add atex_long case to pyplotgen

HingOng committed:
Fine-tune atex_long case with less radical changes

LarsonGroupSysAdmin committed:
Updating the clubb core and silhs version files

HingOng committed:
Merge remote-tracking branch 'upstream/master' into clubb_Coriolis

HingOng commented:
This branch is merged with up-to-date larson-group/clubb_release/master
When l_nontraditional_Coriolis and l_traditional_Coriolis are .false. (default), this merged version outputs additional variables up2_nct, wp2_nct, upwp_nct, upwp_tct, and vpwp_tct to *_zm.nc and wp3_nct and wp2up to *_zt.nc. However, the other output variables remain identical with those generated by larson-group/clubb_release/master for the bomex case.
When l_nontraditional_Coriolis and l_traditional_Coriolis are .true., this merged version outputs an additional flag l_hole_fill = "T". However, the other output variables remain identical with the unmerged version of this branch for the atex_long case (known to be sensitive to the nontraditional Coriolis terms).

vlarson commented:
@HingOng, with the newest version of your branch, do the budgets balance, according to this script?:

https://github.com/larson-group/clubb/blob/master/postprocessing/check_budgets_balance/checkBudget.py

HingOng commented:
> @HingOng, do you happen to have code changes for clubb_intr.F90 in CAM that correspond to these code changes to CLUBB?

Yes, I have them on a local branch.

HingOng commented:
> > > @HingOng, do you happen to have code changes for clubb_intr.F90 in CAM that correspond to these code changes to CLUBB?
> > 
> > 
> > Yes, I have them on a local branch.
> 
> Is the branch on github? Or is there another way to share it?

I just pushed the branch to my fork of CAM:
https://github.com/HingOng/CAM/tree/clubb_Coriolis

Note that this branch is based on an outdated version of CAM and CLUBB.
